src/dmd/traits.d

```diff
@@ -1453,8 +1453,24 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
 
         if (l1 && l2)
         {
-            import core.stdc.string;
+            const(char)* getSerialization(FuncLiteralDeclaration fld)
```

Comment:
Why don't you return `string` here? Comparing D strings is more efficient (no `\0` checking)
You already know the length of the buffer - no need to throw this valuable information away.

Reply:
I know it's easier with strings, but the serialization field is is `const(char)*` and I really haven't seen AST members declared as strings. All of them are `const(char)*`

Reply:
Quoting a conversation with Walter here:

>> On 12/5/2017 11:20 AM, Andrei Alexandrescu wrote:
Hi Walter, Seb noticed that the dmd code is going through unpleasant hoops caused by C-style strings. Do you think incrementally switching to D-style strings would improve the compiler? -- Andrei

> Yes, but we have to be mindful of gdc and ldc both relying on the C++ interface to use dmd, and D strings don't work on that interface.
> I've already converted a bunch of internal stuff to D strings where the interface to the back end was not involved.

(we can even convert the C++ interfaces, but that's significantly more effort)

Reply:
So how do we proceed here? That was my problem also: I could turn the field into a string, but what should I do regarding the header file?

Reply:
better yet: for all the apis used by ldc, gdc
use a compiler dependent (via eg version(LDC)) type Dstring, which for dmd is alias to char[] and for ldc, dmd is alias to extern(C++) struct with size and char* head, along with conversions functions.
that way, dmd code stays clean, safe and keeps length aroud

Reply:
how about this:
```d
extern(C++) Dstring funLib(Dstring input);
```
see full working example here showing using such API from C++ with the function defined in D.

https://github.com/timotheecour/dtools/blob/master/tests/d01_dmd_string/

Reply:
@timotheecour #7893  is a good step in the right direction, we need to start using D strings in the compiler. That would be an effort separated from this.